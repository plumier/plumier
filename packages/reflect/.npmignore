.vscode
coverage
scripts
src 
test
.gitignore
.npmignore
.travis.yml
package-lock.json
tsconfig.build.json
tsconfig.json
yarn-error.log
yarn.lock