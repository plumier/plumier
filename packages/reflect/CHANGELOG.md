# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.1.1](https://github.com/plumier/plumier/compare/v1.1.0...v1.1.1) (2023-01-03)


### Bug Fixes

* **reflect:** Fix issue when reflect async function inside object ([#1033](https://github.com/plumier/plumier/issues/1033)) ([e2be8ca](https://github.com/plumier/plumier/commit/e2be8ca167e9fc3f959d5f6b26d0aa6d8cec49fb))
* **reflect:** Fix issue when reflect function inside object ([#1032](https://github.com/plumier/plumier/issues/1032)) ([719d23a](https://github.com/plumier/plumier/commit/719d23ab41785c02f0e851a2074c86ee0afe6dd1))





# [1.1.0](https://github.com/plumier/plumier/compare/v1.0.6...v1.1.0) (2022-12-19)


### Bug Fixes

* **reflect:** Fix reflect unable to inspect lambda function parameters ([#1030](https://github.com/plumier/plumier/issues/1030)) ([2cc9e82](https://github.com/plumier/plumier/commit/2cc9e82601b70703218da8e09bd38543fb09332e))
* **reflect:** Remove unnecessary base class when create class using property definition ([#1031](https://github.com/plumier/plumier/issues/1031)) ([9134b6d](https://github.com/plumier/plumier/commit/9134b6d787eafaff4ba5a17629b28d5178543055))


### Features

* Add feature to reflect function and lambda function ([#1029](https://github.com/plumier/plumier/issues/1029)) ([f65ec5b](https://github.com/plumier/plumier/commit/f65ec5baa14af9b34793a2ab74f440233f84c6d7))
* **reflect:** Added support for static methods and properties reflection ([#1023](https://github.com/plumier/plumier/issues/1023)) ([f8ee76d](https://github.com/plumier/plumier/commit/f8ee76dbe3a8543f556dfc3d3ac2468376cff482))





## [1.0.6](https://github.com/plumier/plumier/compare/v1.0.5...v1.0.6) (2021-10-04)

**Note:** Version bump only for package @plumier/reflect





## [1.0.5](https://github.com/plumier/plumier/compare/v1.0.4...v1.0.5) (2021-07-25)

**Note:** Version bump only for package @plumier/reflect





## [1.0.3](https://github.com/plumier/plumier/compare/v1.0.2...v1.0.3) (2021-06-28)


### Bug Fixes

* Cache class parser to speed up class introspection ([#970](https://github.com/plumier/plumier/issues/970)) ([aca5751](https://github.com/plumier/plumier/commit/aca5751382bb81bf264df5fbc7dc27df15943874))





## [1.0.2](https://github.com/plumier/plumier/compare/v1.0.0...v1.0.2) (2021-06-06)

**Note:** Version bump only for package @plumier/reflect





## 1.0.1 (2021-05-18)

**Note:** Version bump only for package @plumier/reflect
