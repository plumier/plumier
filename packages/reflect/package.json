{
  "name": "@plumier/reflect",
  "version": "1.0.0-rc.11",
  "description": "Plumier reflection module",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "test": "jest",
    "compile": "tsc -p tsconfig.build.json",
    "clean": "rm -rf src/*.js && rm -rf test/*.js && rm -rf test/**/*.js && rm -rf lib"
  },
  "keywords": [
    "Reflection",
    "Reflect",
    "Metadata"
  ],
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@types/acorn": "4.0.5",
    "acorn": "8.1.1",
    "reflect-metadata": "^0.1.13"
  },
  "devDependencies": {
    "@types/jest": "^26.0.22",
    "@types/node": "^14.14.41",
    "coveralls": "^3.1.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.4",
    "tslib": "^2.2.0",
    "typescript": "^4.2.4"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  }
}
