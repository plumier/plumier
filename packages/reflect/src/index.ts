import "reflect-metadata"

export * from "./types"
export * from "./decorators"
export * from "./helpers"
import { reflect } from "./reflect"
export { reflect }
export default reflect
