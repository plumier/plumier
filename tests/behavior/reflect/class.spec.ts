import { decorateClass, decorateMethod, noop, reflect, type } from "@plumier/reflect"

describe("Class Introspection", () => {
    it("Should inspect class properly", () => {
        class DummyClass {
            dummyMethod() { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect class with constructor parameters", () => {
        class DummyClass {
            constructor(id: number, name: string) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect array on constructor parameters", () => {
        class EmptyClass { }
        @decorateClass({})
        class DummyClass {
            constructor(@reflect.type([EmptyClass]) empty: EmptyClass[]) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect constructor parameters type with decorator", () => {
        @decorateClass({})
        class DummyClass {
            constructor(id: number, name: string) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect class with method parameters", () => {
        class DummyClass {
            dummyMethod(dummy: string, other: any) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect parameters type with method decorator", () => {
        class DummyClass {
            @decorateMethod({})
            dummyMethod(dummy: string, other: any): number { return 1 }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should able to define return type of method", () => {
        class DummyClass {
            @reflect.type([Number])
            method() {
                return [1, 2, 3]
            }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].returnType).toEqual([Number])
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect async method with parameters", () => {
        class DummyClass {
            @decorateMethod({})
            async dummyMethod(dummy: string, other: any): Promise<number> { return 1 }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect destructed parameters type with method decorator", () => {
        class Domain {
            constructor(
                public date: Date,
                public name: string
            ) { }
        }
        class DummyClass {
            @decorateMethod({})
            dummyMethod({ date, name }: Domain): number { return 1 }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0]).toMatchSnapshot()
    })

    it("Should reflect destructed tuple parameters type", () => {
        type Tuple = [string, number]
        class DummyClass {
            @decorateMethod({})
            dummyMethod([name, age]: Tuple) { }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0]).toMatchSnapshot()
    })

    it("Should inspect deep nested destructed method parameter type", () => {
        class Domain {
            constructor(
                public a: Date,
                public b: string,
                public child: ChildDomain,
                public tuple: Tuple
            ) { }
        }
        class ChildDomain {
            constructor(
                public c: Date,
                public d: string,
                public child: GrandChildDomain
            ) { }
        }
        class GrandChildDomain {
            constructor(
                public e: Date,
                public f: string
            ) { }
        }
        type Tuple = [string, Domain]
        class DummyClass {
            @decorateMethod({})
            dummyMethod(@reflect.type(Domain) { a, b, tuple: [name, { a: a1, b: b1 }], child: { c, d, child: { e, f } } }: Domain) { }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0]).toMatchSnapshot()
    })

    it("Should ignore method parameters with this", () => {
        class DummyClass {
            @decorateMethod({})
            dummyMethod(this: string, data: number): number { return 1; }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect array on method parameter but without type override", () => {
        class EmptyClass { }
        class DummyClass {
            @decorateMethod({})
            dummyMethod(empty: EmptyClass[]) { }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0].type).toEqual([Object])
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect array on method parameter", () => {
        class EmptyClass { }
        class DummyClass {
            @decorateMethod({})
            dummyMethod(@reflect.type([EmptyClass]) empty: EmptyClass[]) { }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0].type).toEqual([EmptyClass])
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect array on method parameter using @reflect.type", () => {
        class EmptyClass { }
        class DummyClass {
            @decorateMethod({})
            dummyMethod(@reflect.type([EmptyClass]) empty: EmptyClass[]) { }
        }
        const meta = reflect(DummyClass)
        expect(meta.methods[0].parameters[0].type).toEqual([EmptyClass])
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect destructed method parameter type", () => {
        class Domain {
            constructor(
                public date: Date,
                public name: string
            ) { }
        }
        class DummyClass {
            @decorateMethod({})
            dummyMethod(@reflect.type(Domain) { date, name }: Domain) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect property", () => {
        class DummyClass {
            @reflect.noop()
            myProp: number = 10
            @reflect.noop()
            myOtherProp: string = "hello"
        }

        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect getter and setter", () => {
        class DummyClass {
            get myProp() { return 1 }
            set myProp(val: number) { }
        }

        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect setter only", () => {
        class DummyClass {
            set myProp(val: number) { }
        }

        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect getter and setter with decorator", () => {
        class DummyClass {
            @reflect.noop()
            get myProp() { return 1 }
            set myProp(val: number) { }
        }

        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect parameter properties", () => {
        @reflect.parameterProperties()
        class DummyClass {
            constructor(
                public myProp: number,
                public myOtherProp: string,
            ) { }
        }

        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static method", () => {
        class DummyClass {
            static myStaticMethod() {}
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static method's parameters ", () => {
        class DummyClass {
            static myStaticMethod(par1:number, par2:string) {}
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static method return type with decorator", () => {
        class DummyClass {
            @noop()
            static myStaticMethod(par1:number, par2:string):Number { return 1 }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static method return type with decorator override", () => {
        class DummyClass {
            @type([Number])
            static myStaticMethod(par1:number, par2:string):Number[] { return [1] }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static method's parameter type with decorator", () => {
        class DummyClass {
            @noop()
            static myStaticMethod(par1:number, par2:string) {}
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should able to distinguish between static method and instance method with the same name", () => {
        class DummyClass {
            static method() {}
            method() {}
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static property", () => {
        class DummyClass {
            @noop()
            static myProp:string
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })
    
    it("Should inspect static property type with decorator override", () => {
        class DummyClass {
            @type([Number])
            static myProp:number[]
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static getter and setter", () => {
        class DummyClass {
            static get myProp():number { return 1 }
            static set myProp(value:number) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should inspect static getter and setter type", () => {
        class DummyClass {
            @noop()
            static get myProp():number { return 1 }
            static set myProp(value:number) { }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })
})