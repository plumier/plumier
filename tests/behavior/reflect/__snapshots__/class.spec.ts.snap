// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class Introspection Should able to define return type of method 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          "genericParams": Array [],
          "kind": "Override",
          "target": DummyClass,
          "type": Array [
            Number,
          ],
          Symbol(tinspector:decoratorId): Symbol(override),
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": false,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "method",
      "parameters": Array [],
      "returnType": Array [
        Number,
      ],
      "typeClassification": "Array",
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect array on constructor parameters 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [
          Object {
            "genericParams": Array [],
            "kind": "Override",
            "target": DummyClass,
            "type": Array [
              EmptyClass,
            ],
            Symbol(tinspector:decoratorId): Symbol(override),
            Symbol(tinspector:decoratorOption): Object {
              "allowMultiple": false,
              "applyTo": Array [],
              "inherit": true,
              "removeApplied": true,
            },
          },
        ],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "empty",
        "type": Array [
          EmptyClass,
        ],
        "typeClassification": "Array",
      },
    ],
  },
  "decorators": Array [
    Object {
      Symbol(tinspector:decoratorOption): Object {
        "allowMultiple": true,
        "applyTo": Array [],
        "inherit": true,
        "removeApplied": true,
      },
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect array on method parameter 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "genericParams": Array [],
              "kind": "Override",
              "target": DummyClass,
              "type": Array [
                EmptyClass,
              ],
              Symbol(tinspector:decoratorId): Symbol(override),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": false,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
          ],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "empty",
          "type": Array [
            EmptyClass,
          ],
          "typeClassification": "Array",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect array on method parameter but without type override 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "empty",
          "type": Array [
            Object,
          ],
          "typeClassification": "Array",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect array on method parameter using @reflect.type 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "genericParams": Array [],
              "kind": "Override",
              "target": DummyClass,
              "type": Array [
                EmptyClass,
              ],
              Symbol(tinspector:decoratorId): Symbol(override),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": false,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
          ],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "empty",
          "type": Array [
            EmptyClass,
          ],
          "typeClassification": "Array",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect async method with parameters 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "dummy",
          "type": String,
          "typeClassification": "Primitive",
        },
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 1,
          "kind": "Parameter",
          "name": "other",
          "type": Object,
          "typeClassification": "Primitive",
        },
      ],
      "returnType": Promise,
      "typeClassification": "Class",
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect class properly 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect class with constructor parameters 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "id",
        "type": undefined,
        "typeClassification": undefined,
      },
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 1,
        "kind": "Parameter",
        "name": "name",
        "type": undefined,
        "typeClassification": undefined,
      },
    ],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect class with method parameters 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "dummy",
          "type": undefined,
          "typeClassification": undefined,
        },
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 1,
          "kind": "Parameter",
          "name": "other",
          "type": undefined,
          "typeClassification": undefined,
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect constructor parameters type with decorator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "id",
        "type": Number,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 1,
        "kind": "Parameter",
        "name": "name",
        "type": String,
        "typeClassification": "Primitive",
      },
    ],
  },
  "decorators": Array [
    Object {
      Symbol(tinspector:decoratorOption): Object {
        "allowMultiple": true,
        "applyTo": Array [],
        "inherit": true,
        "removeApplied": true,
      },
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect deep nested destructed method parameter type 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "genericParams": Array [],
              "kind": "Override",
              "target": DummyClass,
              "type": Domain,
              Symbol(tinspector:decoratorId): Symbol(override),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": false,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
          ],
          "fields": Array [
            "a",
            "b",
            Object {
              "child": Array [
                "c",
                "d",
                Object {
                  "child": Array [
                    "e",
                    "f",
                  ],
                },
              ],
            },
          ],
          "index": 0,
          "kind": "Parameter",
          "name": "{ a, b, child: { c, d, child: { e, f } } }",
          "type": Domain,
          "typeClassification": "Class",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect destructed method parameter type 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "genericParams": Array [],
              "kind": "Override",
              "target": DummyClass,
              "type": Domain,
              Symbol(tinspector:decoratorId): Symbol(override),
              Symbol(tinspector:decoratorOption): Object {
                "allowMultiple": false,
                "applyTo": Array [],
                "inherit": true,
                "removeApplied": true,
              },
            },
          ],
          "fields": Array [
            "date",
            "name",
          ],
          "index": 0,
          "kind": "Parameter",
          "name": "{ date, name }",
          "type": Domain,
          "typeClassification": "Class",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect destructed parameters type with method decorator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "par",
          "type": String,
          "typeClassification": "Primitive",
        },
        Object {
          "decorators": Array [],
          "fields": Array [
            "date",
            "name",
          ],
          "index": 1,
          "kind": "Parameter",
          "name": "{ date, name }",
          "type": Domain,
          "typeClassification": "Class",
        },
      ],
      "returnType": Number,
      "typeClassification": "Primitive",
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect getter and setter 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": get myProp,
      "kind": "Property",
      "name": "myProp",
      "set": set myProp,
      "type": undefined,
      "typeClassification": undefined,
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect getter and setter with decorator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [
        Object {
          "kind": "Noop",
          "target": DummyClass,
          Symbol(tinspector:decoratorId): Symbol(noop),
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": false,
            "applyTo": Array [],
            "inherit": false,
            "removeApplied": true,
          },
        },
      ],
      "get": get myProp,
      "kind": "Property",
      "name": "myProp",
      "set": set myProp,
      "type": Number,
      "typeClassification": "Primitive",
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect parameter properties 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "myProp",
        "type": Number,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 1,
        "kind": "Parameter",
        "name": "myOtherProp",
        "type": String,
        "typeClassification": "Primitive",
      },
    ],
  },
  "decorators": Array [
    Object {
      "type": "ParameterProperties",
      Symbol(tinspector:decoratorId): Symbol(paramProp),
      Symbol(tinspector:decoratorOption): Object {
        "allowMultiple": false,
        "applyTo": Array [],
        "inherit": true,
        "removeApplied": true,
      },
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 0,
      "isParameter": true,
      "kind": "Property",
      "name": "myProp",
      "set": undefined,
      "type": Number,
      "typeClassification": "Primitive",
    },
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 1,
      "isParameter": true,
      "kind": "Property",
      "name": "myOtherProp",
      "set": undefined,
      "type": String,
      "typeClassification": "Primitive",
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect parameters type with method decorator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": true,
            "applyTo": Array [],
            "inherit": true,
            "removeApplied": true,
          },
        },
      ],
      "kind": "Method",
      "name": "dummyMethod",
      "parameters": Array [
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 0,
          "kind": "Parameter",
          "name": "dummy",
          "type": String,
          "typeClassification": "Primitive",
        },
        Object {
          "decorators": Array [],
          "fields": Object {},
          "index": 1,
          "kind": "Parameter",
          "name": "other",
          "type": Object,
          "typeClassification": "Primitive",
        },
      ],
      "returnType": Number,
      "typeClassification": "Primitive",
    },
  ],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect property 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [
        Object {
          "kind": "Noop",
          "target": DummyClass,
          Symbol(tinspector:decoratorId): Symbol(noop),
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": false,
            "applyTo": Array [],
            "inherit": false,
            "removeApplied": true,
          },
        },
      ],
      "get": undefined,
      "kind": "Property",
      "name": "myProp",
      "set": undefined,
      "type": Number,
      "typeClassification": "Primitive",
    },
    Object {
      "decorators": Array [
        Object {
          "kind": "Noop",
          "target": DummyClass,
          Symbol(tinspector:decoratorId): Symbol(noop),
          Symbol(tinspector:decoratorOption): Object {
            "allowMultiple": false,
            "applyTo": Array [],
            "inherit": false,
            "removeApplied": true,
          },
        },
      ],
      "get": undefined,
      "kind": "Property",
      "name": "myOtherProp",
      "set": undefined,
      "type": String,
      "typeClassification": "Primitive",
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Class Introspection Should inspect setter only 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": undefined,
      "kind": "Property",
      "name": "myProp",
      "set": set myProp,
      "type": undefined,
      "typeClassification": undefined,
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;
